### PR TITLE
feat(oracle): update params

### PR DESCRIPTION
**Pull request by leonz789**
## Description

Implement service for updating params

----

Closes #XXX

## Summary by CodeRabbit

- **New Features**
  - Introduced a new command to update parameters via the CLI.
  - Added support for managing chains, tokens, sources, rules, and token feeders within the parameters.

- **Bug Fixes**
  - Fixed various parameter handling issues and improved validation logic.

- **Tests**
  - Added comprehensive test cases for updating parameters, including chains, sources, tokens, and token feeders.

- **Error Handling**
  - Introduced new error messages for better clarity on specific issues (e.g., getting prices and invalid parameters).

### CODE REVIEW

**Review by MaxMustermann2 — APPROVED** [review]
Please add comments for the functions which validate the `struct` members of `Params`

**Comment by leonz789** [comment]
> Please add comments for the functions which validate the `struct` members of `Params`

updated

**Comment by cloud8little** [comment]
**make build** error:

```
x/oracle/types/params.go:124:10: undefined: ErrInvalidParams
x/oracle/types/params.go:138:11: undefined: ErrInvalidParams
x/oracle/types/params.go:142:11: undefined: ErrInvalidParams
x/oracle/types/params.go:146:11: undefined: ErrInvalidParams
x/oracle/types/params.go:150:11: undefined: ErrInvalidParams
x/oracle/types/params.go:176:11: undefined: ErrInvalidParams
x/oracle/types/params.go:191:12: undefined: ErrInvalidParams
x/oracle/types/params.go:197:13: undefined: ErrInvalidParams
x/oracle/types/params.go:222:14: undefined: ErrInvalidParams
x/oracle/types/params.go:225:14: undefined: ErrInvalidParams
x/oracle/types/params.go:225:14: too many errors
make: *** [Makefile:134: build] Error 1
```

File: x/oracle/types/message_update_params.go [hunk 1]
```diff
@@ -25,15 +26,27 @@ func (msg *MsgUpdateParams) GetSignBytes() []byte {
...
 // GetSigners returns the expected signers for a MsgUpdateParams message
 func (msg *MsgUpdateParams) GetSigners() []sdk.AccAddress {
 	addr, _ := sdk.AccAddressFromBech32(msg.Authority)
 	return []sdk.AccAddress{addr}
 }
+
+func NewMsgUpdateParams(creator, paramsJSON string) *MsgUpdateParams {
+	var p Params
+	if err := json.Unmarshal([]byte(paramsJSON), &p); err != nil {
+		panic("invalid json for parmas")
```
Comment (cloud8little) [inline x/oracle/types/message_update_params.go]:
```suggestion
		panic("invalid json for params")
```

Comment (leonz789) [reply x/oracle/types/message_update_params.go]: updated

**Comment by cloud8little** [comment]
Start a node initialize with oracle params of token_id = 1 with following values:

`exocored q oracle params`
```
  - end_block: "100"
    interval: "10"
    rule_id: "1"
    start_base_block: "20"
    start_round_id: "1"
    token_id: "1"
```
note that the end_block is set as 100, at the height of 10, trying to update end_block to 200, but found that other fields are all set as default 0.

`exocored tx oracle update-params '{"token_feeders":[{"token_id":"1", "end_block":"200"}]}' --from local_funded_account --home /root/.test-pr83 --gas-prices 90395116aexo`
```
chains: []
tokens: []
sources: []
rules: []
tokenfeeders:
- tokenid: 1
  ruleid: 0
  startroundid: 0
  startbaseblock: 0
  interval: 0
  endblock: 200
maxnonce: 0
thresholda: 0
thresholdb: 0
mode: 0
maxdetid: 0
```

File: x/oracle/keeper/params_test.go [hunk 2]
```diff
@@ -16,3 +16,117 @@ func TestGetParams(t *testing.T) {
...
+			p, err := p.UpdateTokenFeeder(&tt.tokenFeeder, tt.height)
+			if err == nil {
+				err = p.Validate()
+			}
+			if tt.err != nil {
+				require.ErrorIs(t, err, tt.err)
+			}
+		})
+	}
+}
```
Comment (cloud8little) [inline x/oracle/keeper/params_test.go:217]: could you add UT for update end_block for different conditions?

Comment (leonz789) [reply x/oracle/keeper/params_test.go:217]: there're cases for end_block update:
1. history block
2. (endblock-startblock)%interval>=maxNonce

Comment (leonz789) [reply x/oracle/keeper/params_test.go:217]: > could you add UT for update end_block for different conditions.

add some new cases:
1. endblock<=startbasedblock
2. for existing not started feeder, set endblock to history
3. for existing started feeder, set endblock to history

File: x/oracle/keeper/params_test.go [hunk 3]
```diff
@@ -92,6 +92,35 @@ func TestUpdateTokenFeeder(t *testing.T) {
...
+			err:    types.ErrInvalidParams.Wrapf("invalid tokenFeeder to update, invalid EndBlock, currentHeight: %d, set: %d", 6, 5),
+		},
+		{
+			name: "invalid update, for existing feeder, feeder started, and set endblock to history height including 0",
+			tokenFeeder: types.TokenFeeder{
+				TokenID: 2,
+			},
+			height: 11,
+			err:    types.ErrInvalidParams.Wrapf("invalid tokenFeeder to update, invalid EndBlock, currentHeight: %d, set: %d", 11, 0),
+		},
```
Comment (cloud8little) [inline x/oracle/keeper/params_test.go:112]: add positive cases for startbaseblock and endblock plz.  also add negative cases when trying to update end_block when (endBlock-startBasedBlock)%interval<maxNonce. 

1. endBlock is a future block
2. endBlock > startBasedBlock
3. (endBlock-startBasedBlock)%interval≥maxNonce

Comment (leonz789) [reply x/oracle/keeper/params_test.go:112]: `(endblock-startBaseBlock) % interval < maxNonce` had been tested in this file('not set properly')

Comment (leonz789) [reply x/oracle/keeper/params_test.go:112]: add positive cases

**Comment by cloud8little** [comment]
1. Start a single validator with oracle default params. 
2. Add a new token feeder without adding a new token.
3. Transaction failed with invalid EndBlock message.

```
raw_log: 'failed to execute message; message index: 0: invalid tokenFeeder, invalid
  EndBlock: invalid params'
```

**Review by cloud8little — APPROVED** [review]

File: x/oracle/types/params.go [hunk 4]
```diff
@@ -24,7 +24,23 @@ func ParamKeyTable() paramtypes.KeyTable {
 
 // NewParams creates a new Params instance
 func NewParams() Params {
-	return Params{}
+	return Params{
+		// maximum number of transactions be submitted in one round from a validator
+		MaxNonce: 1,
```
Comment (bwhour) [inline x/oracle/types/params.go:29]: > Change MaxNonce to **MaxTransactionPerRound** for clearer expression of its meaning. 
> 
> Change Mode to **PriceConfirmationMode** for a more explicit indication of its purpose.

File: x/oracle/types/params.go [hunk 5]
```diff
@@ -97,15 +117,340 @@ func (p *Params) ParamSetPairs() paramtypes.ParamSetPairs {
 
 // Validate validates the set of params
 func (p Params) Validate() error {
```
Comment (bwhour) [inline x/oracle/types/params.go:120]: > Add TODO: **Validate function** Performance may be poor when dealing with large amounts of data: consider using parallel processing to improve performance, such as  parallel verification of Chains, Tokens, Rules, and Sources in the Validate function.

**Review by bwhour — APPROVED** [review]

File: x/oracle/types/params.go (hunk 4 above)
Comment (leonz789) [reply x/oracle/types/params.go:29]: We have this 'nonce' field set in message structure, and this MaxNonce is the limitation for that nonce, it also works like `nonce` in transaction but in this special case of oracle to distinguish transactions.
For mode it's not only confirmation(threshold or all messages received) but also how to update price for single validator with new price for deterministic/non-deteministic sources. We've go detailed comments for it, maybe don't put all details in the field name.

File: x/oracle/types/params.go (hunk 5 above)
Comment (leonz789) [reply x/oracle/types/params.go:120]: updated, will do this in later patch

**Pull request merged**